libs/cosmos-sdk/store/mpt/snapshot.go

```diff
@@ -0,0 +1,206 @@
+package mpt
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/state/snapshot"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+	snap "github.com/okx/okbchain/libs/cosmos-sdk/store/mpt/snapshot"
+	mpttypes "github.com/okx/okbchain/libs/cosmos-sdk/store/mpt/types"
+)
+
+var (
+	gDisableSnapshot = false
+	gSnapshotRebuild = false
+)
+
+const (
+	snapshotMemoryLayerCount = 1
+)
+
+func DisableSnapshot() {
+	gDisableSnapshot = true
+}
+
+func SetSnapshotRebuild(rebuild bool) {
+	gSnapshotRebuild = rebuild
+}
+
+func (ms *MptStore) openSnapshot() error {
+	if ms == nil || ms.db == nil || ms.trie == nil || snap.GetDiskDB() == nil || gDisableSnapshot {
+		return fmt.Errorf("mpt store is nil or mpt trie is nil")
+	}
+	// If the chain was rewound past the snapshot persistent layer (causing
+	// a recovery block number to be persisted to disk), check if we're still
+	// in recovery mode and in that case, don't invalidate the snapshot on a
+	// head mismatch.
+	var recovery bool
+
+	version := ms.CurrentVersion()
+	if layer := rawdb.ReadSnapshotRecoveryNumber(snap.GetDiskDB()); layer != nil && *layer > uint64(version) {
```

Comment:
use `InstanceOfMptStore().TrieDB().DiskDB()` instead of `snap.GetDiskDB()`

Reply:
fixed